store/src/errors.rs

```diff
@@ -1,64 +1,154 @@
-use miden_crypto::hash::rpo::RpoDigest;
-use miden_node_proto::error::ParseError;
+use std::io;
 
-#[derive(Clone, Debug)]
+use deadpool_sqlite::PoolError;
+use miden_crypto::{
+    hash::rpo::RpoDigest,
+    merkle::{MerkleError, MmrError},
+    utils::DeserializationError,
+};
+use miden_node_proto::{block_header::BlockHeader, errors::ParseError};
+use prost::DecodeError;
+use rusqlite::types::FromSqlError;
+use thiserror::Error;
+use tokio::sync::oneshot::error::RecvError;
+
+#[derive(Debug, Error)]
+pub enum DbError {
+    #[error("Missing database connection: {0}")]
+    MissingDbConnection(#[from] PoolError),
+    #[error("SQLite error: {0}")]
+    SqliteError(#[from] rusqlite::Error),
+    #[error("SQLite error: {0}")]
+    FromSqlError(#[from] FromSqlError),
+    #[error("SQLite migration error: {0}")]
+    SqliteMigrationError(#[from] rusqlite_migration::Error),
+    #[error("Pool build error: {0}")]
+    PoolBuildError(#[from] deadpool_sqlite::BuildError),
+    #[error("I/O error: {0}")]
+    IoError(#[from] io::Error),
+    #[error("Prost decode error: {0}")]
+    DecodeError(#[from] DecodeError),
+    #[error("SQLite pool interaction task failed: {0}")]
+    InteractionTaskError(#[from] InteractionTaskError),
+    #[error("Genesis block error: {0}")]
+    GenesisBlockError(#[from] GenesisBlockError),
+    #[error("State error: {0}")]
+    StateError(Box<StateError>),
+    #[error("Block database is empty")]
+    BlockDbIsEmpty,
+    #[error("Decoding nullifier from database failed: {0}")]
+    NullifierDecodingError(DeserializationError),
+    #[error("Block applying was broken because of closed channel on state side: {0}")]
+    BlockApplyingBrokenBecauseOfClosedChannel(RecvError),
+}
+
+impl From<StateError> for DbError {
+    fn from(value: StateError) -> Self {
+        Self::StateError(Box::new(value))
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum InteractionTaskError {
+    #[error("Migration task failed: {0}")]
+    MigrationTaskFailed(String),
+    #[error("Select nullifiers task failed: {0}")]
+    SelectNullifiersTaskFailed(String),
+    #[error("Select notes task failed: {0}")]
+    SelectNotesTaskFailed(String),
+    #[error("Select accounts task failed: {0}")]
+    SelectAccountsTaskFailed(String),
+    #[error("Select block header task failed: {0}")]
+    SelectBlockHeaderTaskFailed(String),
+    #[error("Select block headers task failed: {0}")]
+    SelectBlockHeadersTaskFailed(String),
+    #[error("Select account hashes task failed: {0}")]
+    SelectAccountHashesTaskFailed(String),
+    #[error("Get state sync task failed: {0}")]
+    GetStateSyncTaskFailed(String),
+    #[error("Apply block task failed: {0}")]
+    ApplyBlockTaskFailed(String),
+}
+
+#[derive(Debug, Error)]
+pub enum GenesisBlockError {
```

Comment:
nit: I would rename this into just `GenesisError`.

Also, let's implement `Clone` an `PartialEq` here and for all other error types.

Reply:
> Also, let's implement `Clone` an `PartialEq` here and for all other error types.

I think we can implement `PartialEq`, but will have difficulties with implementing `Clone`, because of underlying error types, which don't implement it. Do we really need `Clone` for errors?

Reply:
I've checked, `PartialEq` also wasn't implemented in the most of underlying error types.